apps/lib/apps.c

```diff
@@ -2498,74 +2503,37 @@ BIO *bio_open_owner(const char *filename, int format, int private)
      * context is.
      */
     if (!textmode)
-        fd = open(filename, mode, 0600, "ctx=bin");
+        fd = open(filename, flags, 0600, "ctx=bin");
     else
 #endif
-        fd = open(filename, mode, 0600);
-    if (fd < 0)
-        goto err;
-    fp = fdopen(fd, modestr('w', format));
-    if (fp == NULL)
-        goto err;
-    bflags = BIO_CLOSE;
-    if (textmode)
-        bflags |= BIO_FP_TEXT;
-    b = BIO_new_fp(fp, bflags);
-    if (b)
-        return b;
-
- err:
-    BIO_printf(bio_err, "%s: Can't open \"%s\" for writing, %s\n",
-               opt_getprog(), filename, strerror(errno));
-    ERR_print_errors(bio_err);
-    /* If we have fp, then fdopen took over fd, so don't close both. */
-    if (fp)
-        fclose(fp);
-    else if (fd >= 0)
-        close(fd);
-    return NULL;
-}
-
-static BIO *bio_open_default_(const char *filename, char mode, int format,
-                              int quiet)
-{
-    BIO *ret;
-
-    if (filename == NULL || strcmp(filename, "-") == 0) {
-        ret = mode == 'r' ? dup_bio_in(format) : dup_bio_out(format);
-        if (quiet) {
-            ERR_clear_error();
-            return ret;
-        }
-        if (ret != NULL)
-            return ret;
-        BIO_printf(bio_err,
-                   "Can't open %s, %s\n",
-                   mode == 'r' ? "stdin" : "stdout", strerror(errno));
-    } else {
-        ret = BIO_new_file(filename, modestr(mode, format));
-        if (quiet) {
-            ERR_clear_error();
-            return ret;
-        }
-        if (ret != NULL)
-            return ret;
-        BIO_printf(bio_err,
-                   "Can't open %s for %s, %s\n",
-                   filename, modeverb(mode), strerror(errno));
+        fd = open(filename, flags, 0600);
+    if (fd < 0
+            || (fp = fdopen(fd, modestr('w', format))) == NULL
+            || (b = BIO_new_fp(fp, bflags)) == NULL) {
+        BIO_printf(bio_err, "%s: Can't open \"%s\" for writing, %s\n",
+                   opt_getprog(), filename, strerror(errno));
+        ERR_print_errors(bio_err);
+        if (fp != NULL)
+            fclose(fp);
+        else if (fd >= 0)
+            close(fd);
+        exit(1);
     }
-    ERR_print_errors(bio_err);
-    return NULL;
+    return b;
 }
 
 BIO *bio_open_default(const char *filename, char mode, int format)
 {
-    return bio_open_default_(filename, mode, format, 0);
-}
+    BIO *b;
 
-BIO *bio_open_default_quiet(const char *filename, char mode, int format)
-{
-    return bio_open_default_(filename, mode, format, 1);
+    /* Using stdin or stdout? */
+    if (filename == NULL || strcmp(filename, "-") == 0)
+        return mode == 'r' ? dup_bio_in(format) : dup_bio_out(format);
+
+    b = BIO_new_file(filename, modestr(mode, format));
+    if (mode == 'r')
+        return check_bio(b, "stdin", format);
+    return check_bio(b, "stdout", format);
```

Comment:
Here, you pass `format` to `check_bio`.  Why?

Reply:
I was also complaining about the mixup of format, file mode, and BIO flags values, but @richsalz  did not want to fix this as part of this PR.
Yet actually this PR would be a better place to do it correctly right away.

Reply:
> Here, you pass `format` to `check_bio`. Why?

This issue has been postponed - see #12056.

Reply:
Because ```format``` is what was used in the BIO function.